ArcGISRuntimeSDKQt_CppSamples/Layers/ApplyMosaicRuleToRasters/ApplyMosaicRuleToRasters.cpp

```diff
@@ -0,0 +1,125 @@
+#include "RasterLayer.h"
+
+using namespace Esri::ArcGISRuntime;
+
+ApplyMosaicRuleToRasters::ApplyMosaicRuleToRasters(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_map(new Map(Basemap::topographicVector(this), this)),
+  m_mosaicRule(new MosaicRule(this))
+{
+}
+
+ApplyMosaicRuleToRasters::~ApplyMosaicRuleToRasters() = default;
+
+void ApplyMosaicRuleToRasters::init()
+{
+  // Register the map view for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<ApplyMosaicRuleToRasters>("Esri.Samples", 1, 0, "ApplyMosaicRuleToRastersSample");
+}
+
+MapQuickView* ApplyMosaicRuleToRasters::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void ApplyMosaicRuleToRasters::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  // Create image service raster from image server
+  m_imageServiceRaster = new ImageServiceRaster(QUrl("https://sampleserver7.arcgisonline.com/arcgis/rest/services/amberg_germany/ImageServer"), this);
+
+  // Set Mosaic Rule if none exists on the image service raster
+  if (!m_imageServiceRaster->mosaicRule())
+    m_imageServiceRaster->setMosaicRule(m_mosaicRule);
+
+  // Create a raster layer form the image service raster
+  m_rasterLayer = new RasterLayer(m_imageServiceRaster, this);
+
+  // Once loaded change the viewpoint the the center of the raster layers full extent
+  connect(m_rasterLayer, &RasterLayer::doneLoading, this, [this](Error e)
+  {
+    if (e.isEmpty())
+    {
+      m_rasterLoaded = true;
+      emit rasterLoadedChanged();
+      m_mapView->setViewpointCenter(m_rasterLayer->fullExtent().center(), 25000.0);
+    }
+  });
+
+  m_map->operationalLayers()->append(m_rasterLayer);
+  emit mapViewChanged();
+}
+
+void ApplyMosaicRuleToRasters::applyRasterRule(const QString& ruleString)
+{
+  // Reset to clear previous mosaic rule parameters
+  ApplyMosaicRuleToRasters::resetMosaicRule();
+
+  if (ruleString == "None")
+  {
+    m_mosaicRule->setMosaicMethod(MosaicMethod::None);
+  }
+  else if (ruleString == "NorthWest")
+  {
+    m_mosaicRule->setMosaicMethod(MosaicMethod::Northwest);
+    m_mosaicRule->setMosaicOperation(MosaicOperation::First);
+  }
+  else if (ruleString == "Center")
+  {
+    m_mosaicRule->setMosaicMethod(MosaicMethod::Center);
+    m_mosaicRule->setMosaicOperation(MosaicOperation::Blend);
+  }
+  else if (ruleString == "ByAttribute")
+  {
+    m_mosaicRule->setMosaicMethod(MosaicMethod::Attribute);
+    m_mosaicRule->setSortField("OBJECTID");
+
+  }
+  else if (ruleString  == "LockRaster")
+  {
+    m_mosaicRule->setMosaicMethod(MosaicMethod::LockRaster);
+    m_mosaicRule->setLockRasterIds(QList<qint64>{1,7,12});
+  }
+  m_imageServiceRaster->setMosaicRule(m_mosaicRule);
+}
+
+// Helper function to reset the mosaic rule
+void ApplyMosaicRuleToRasters::resetMosaicRule()
+{
+  delete m_mosaicRule;
+  m_mosaicRule = nullptr;
```

Comment:
How about creating the MosaicRule by value when setting the parameters so once it goes out of scope the system cleans it up? Then it would remove the need for that helper function as well as remove a member variable. I pushed the jist of the change so you can see it.

Reply:
>How about creating the MosaicRule by value...

No. This can lead to assumptions about how the API internals work.

```c++
m_imageServiceRaster->setMosaicRule(&myMosaicRule);
```

Since that value object will get destructed on the next line, you can make the case it is similar to doing this, and there is no expectation this would work:
```c++
myMosaicRule = new MosaicRule(this);
...
m_imageServiceRaster->setMosaicRule(myMosaicRule);
delete myMosaicRule;
// is this the same as m_imageServiceRaster->setMosaicRule(nullptr); ?
```

Reply:
That's a good point. I reverted that last commit to go back to what it was originally. Thank you for the input!